### PR TITLE
Enclosing name case

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR forbids data members in classes, structs, and exceptions from having the same name as their enclosing container (case-insensitively)

### CODE REVIEW

File: cpp/src/Slice/Parser.cpp [hunk 1]
```diff
@@ -2503,6 +2503,22 @@ Slice::ClassDef::createDataMember(const string& name, const TypePtr& type, bool
         return nullptr;
     }
 
+    // Check whether enclosing class has the same name.
+    if (name == this->name())
+    {
+        _unit->error("data member `" + name + "' cannot share the same name as its enclosing class");
```
Comment (bernardnormier) [inline cpp/src/Slice/Parser.cpp:2509]: I am not fond of "cannot share the same name": cannot have the same name?

**Comment by InsertCreativityHere** [comment]
Operations were already forbidden from sharing the same name as their containing interface before my PR.
But it's still allowable for modules and their contents to share names, would this also be a problem in C#?

File: cpp/src/Slice/Parser.cpp (hunk 1 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Parser.cpp:2509]: Alright! It can be done!

**Comment by bernardnormier** [comment]
> But it's still allowable for modules and their contents to share names, would this also be a problem in C#?
That's all fine in C# (with namespaces). So we use the same rules in Slice.

File: cpp/src/Slice/Parser.cpp (hunk 1 above)
Comment (pepone) [reply cpp/src/Slice/Parser.cpp]: or maybe

```
        _unit->error("`" + name + "': member names cannot be the same as their enclosing type");
```

File: cpp/src/Slice/Parser.cpp [hunk 2]
```diff
@@ -2503,6 +2503,22 @@ Slice::ClassDef::createDataMember(const string& name, const TypePtr& type, bool
...
+    }
+
+    if (ciequals(name, this->name()))
+    {
+        _unit->error("data member `" + name + "' differs only in capitalization from its enclosing class named `"
+                     + this->name() + "'");
+        _unit->note(this, "class `" + this->name() + "' is defined here");
+        return nullptr;
+    }
+
```
Comment (pepone) [inline cpp/src/Slice/Parser.cpp:2521]: Maybe we can move this check to a common method

Comment (InsertCreativityHere) [reply cpp/src/Slice/Parser.cpp:2521]: I have in my next PR that unifies DataMember and ParamDecl. There's loads of logic that is shared between Struct, Class, Exception, and even somewhat Interface that I brought into common methods, this check included.

File: cpp/src/Slice/Parser.cpp (hunk 1 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Parser.cpp]: I prefer the more information dense error messages, even if some of the information is ancillary. What do you think Bernard? Any preference to message format?

**Pull request merged**